File: ios/App/Settings/SettingsRootViewModel.swift
// ---------------------------------------------------------------------------
// SettingsRootViewModel — the thin state-holder for the root Settings category
// list. Its ONE job is the Admin-group gate: it folds the KMP
// ObserveSettingsAccessUseCase (me.isAdmin + fish flag) into a published access
// state the root view reads. Everything else on the root page (category rows,
// logout, update footer) is static or bound to the shared UpdateModel — no VM.
//
// One-shot load (the usecase is a `suspend` one-shot, not a Flow): the view calls
// `load()` from `.task`; a route re-entry rebuilds the VM and reloads. The result
// envelope is folded EXHAUSTIVELY (success / failure / loading) per the mobile-data
// result-envelope rule — a failure or still-loading access degrades to "no Admin
// group", never a crash and never a leaked admin surface.
// ---------------------------------------------------------------------------
import Foundation
import MobileData

@MainActor
@Observable
final class SettingsRootViewModel {
    /// Access-gate state for the root page. Admin surfaces only in `.ready(isAdmin: true)`.
    enum AccessState: Equatable {
        case loading
        case ready(isAdmin: Bool, fishBrowseEnabled: Bool)
        case failed
    }

    private(set) var access: AccessState = .loading

    private let observeAccess: ObserveSettingsAccessUseCase
    private let log = AppLog("settings", "root-vm")

    init(observeAccess: ObserveSettingsAccessUseCase) {
        self.observeAccess = observeAccess
    }

    /// True only once access has resolved AND the user is an admin. Loading / failed
    /// both read false, so the Admin group is hidden by default (fail-safe).
    var showAdmin: Bool {
        if case .ready(let isAdmin, _) = access { return isAdmin }
        return false
    }

    /// Resolve the access flags. Idempotent; safe to call on each `.task`. Folds the
    /// SentientResult envelope exhaustively; a thrown error (SKIE suspend-bridge /
    /// cancellation) degrades to `.failed` (or is ignored on cancellation).
    func load() async {
        log.info("access.load")
        do {
            let result = try await observeAccess.invoke()
            switch onEnum(of: result) {
            case .success(let s):
                access = .ready(isAdmin: s.data.isAdmin, fishBrowseEnabled: s.data.fishBrowseEnabled)
                log.info("access.ready isAdmin=\(s.data.isAdmin) fish=\(s.data.fishBrowseEnabled)")
            case .failure:
                access = .failed
                log.warn("access.failed")
            case .loading:
                access = .loading
            }
        } catch is CancellationError {
            // View disappeared / task replaced — not a real failure.
        } catch {
            access = .failed
            log.warn("access.threw reason=\(error.localizedDescription)")
        }
    }
}

File: android/src/main/kotlin/io/sentient/android/settings/SettingsRootViewModel.kt
// ---------------------------------------------------------------------------
// SettingsRootViewModel — the thin state-holder for the root Settings list.
//
// Its only job is to resolve the access flags (isAdmin, fishBrowseEnabled) so the
// root list can gate the Admin group. It collects the single combine usecase
// (ObserveSettingsAccessUseCase) resolved from the connection scope; the fold /
// degrade rules live in that usecase, not here (thin VM per the architecture rule).
// ---------------------------------------------------------------------------
package io.sentient.android.settings

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import io.sentient.mobiledata.result.SentientResult
import io.sentient.mobiledata.usecase.settings.ObserveSettingsAccessUseCase
import io.sentient.mobiledata.usecase.settings.SettingsAccess
import io.sentient.mobilesdk.log.createLogger
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch

/**
 * Drives the root Settings list's access gate. [observeAccess] is resolved from the
 * connection-scoped SettingsComponent. [access] is null while loading; a failure
 * leaves it null (the Admin group simply stays hidden — never blocks the list).
 */
class SettingsRootViewModel(
    private val observeAccess: ObserveSettingsAccessUseCase,
) : ViewModel() {
    private val log = createLogger("android", "settings-root-viewmodel")

    private val _access = MutableStateFlow<SettingsAccess?>(null)

    /** Root-page access flags; null until the first resolve (or on failure). */
    val access: StateFlow<SettingsAccess?> = _access.asStateFlow()

    init {
        refresh()
    }

    /** (Re)load the access flags. Never throws — the usecase returns a typed result. */
    fun refresh() {
        viewModelScope.launch {
            when (val r = observeAccess()) {
                is SentientResult.Success -> {
                    _access.value = r.data
                    log.info("access.loaded", mapOf("isAdmin" to r.data.isAdmin))
                }
                is SentientResult.Failure -> log.warn("access.failed", mapOf("kind" to r.error.kind))
                is SentientResult.Loading -> Unit
            }
        }
    }
}
